### ISSUE TITLE
Multiple iptables INPUT entries after a while

### ISSUE THREAD

**Issue by TymurGubayev** [dlenski/vpn-slice#69]
I'm using vpn-slice 0.15 like follows.

```
$ sudo cat /etc/systemd/system/vpn.client.service
[Unit]
Description=Connect to Client VPN
After=network.target

[Service]
Type=simple
Environment=password=abc123
ExecStart=/bin/sh -c 'echo $password | sudo openconnect --passwd-on-stdin -s \'vpn-slice 123.123.123.0/24\' --interface=tunC --protocol=gp --user=vpnuser --usergroup=gateway --servercert pin-sha256:<some fingerprint> vpn.client.com'
Restart=always
RestartSec=60

[Install]
WantedBy=multi-user.target
```
Please notice: it's **openconnect** with protocol **gp**.

And it works perfectly and does exactly what I need... But there is one small nuisance that's bothering me.

After some time the `iptables` accumulates entries like following.

```
$sudo iptables -vnL --line-numbers
...
4        0     0 DROP       all  --  tunC *       0.0.0.0/0            0.0.0.0/0
5        0     0 ACCEPT     all  --  tunC *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
6        0     0 DROP       all  --  tunC *       0.0.0.0/0            0.0.0.0/0
7        0     0 ACCEPT     all  --  tunC *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
```

I think this happens when other side resets the connection for whatever reason.

**Comment by dlenski** [comment]
> After some time the `iptables` accumulates entries like following.

Hrmmm. I have never seen this phenomenon before. I actually didn't even know that `iptables` would _allow_ multiple _identical_ entries to be created.  (What Linux kernel and distro are you running?)

> I think this happens when other side resets the connection for whatever reason.

What do you mean by "other side" and "the connection" here? Are you saying that you think this comes from the OpenConnect process (`openconnect --version`?) losing its connection to the VPN gateway and then reconnecting.

That _should not_ happen, because `vpn-slice` doesn't actually do anything when it is called with `reason=attempt-reconnect` or `reason=reconnect`.

> ```
> $ sudo cat /etc/systemd/system/vpn.client.service
> ```

Is it possible that OpenConnect is being uncleanly terminated by systemd? If vpn-slice is never getting called with `reason=disconnect`, but the VPN is getting restarted by systemd, that could well produce this effect.

**Comment by TymurGubayev** [comment]
Linux version:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster
$ cat /proc/version
Linux version 4.19.0-9-amd64 (debian-kernel@lists.debian.org) (gcc version 8.3.0 (Debian 8.3.0-6)) #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07)
```

> What do you mean by "other side" and "the connection" here?

Other side: the VPN server (`vpn.client.com`). The connection: my VPN connection to it using the service described above.

Excerpt from `journalctl -r -u vpn.client.service` aroung what *I think* triggers the issue.
```
Dec 22 11:43:32 debian-router sh[2263]: SSL negotiation with vpn.client.com
Dec 22 11:43:32 debian-router sh[2263]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 22 11:43:32 debian-router sh[2263]: Packet receive error: Input/output error
Dec 22 11:43:32 debian-router sh[2263]: SSL read error: Error in the pull function.; reconnecting.
Dec 22 09:59:24 debian-router sh[2263]: Connected as 123.123.123.123, using SSL, with ESP disabled
Dec 22 09:59:24 debian-router sh[2263]: Set up UDP failed; using SSL instead
```

```
$ sudo openconnect --version
OpenConnect version v8.02-1+deb10u1
Using GnuTLS. Features present: TPMv2, PKCS#11, RSA software token, HOTP software token, TOTP software token, Yubikey OATH, System keys, DTLS, ESP
Supported protocols: anyconnect (default), nc, gp
```

> Is it possible that OpenConnect is being uncleanly terminated by systemd?

I'm not sure how would I check this.

**Comment by dlenski** [comment]
> Excerpt from `journalctl -r -u vpn.client.service` aroung what I think triggers the issue.

A reconnection without termination of the openconnect process _should not_ cause this issue for the reason described [above](https://github.com/dlenski/vpn-slice/issues/69#issuecomment-749697737).

> > Is it possible that OpenConnect is being uncleanly terminated by systemd?
>
> I'm not sure how would I check this.

The default signal that systemd sends when stopping or restarting is `SIGTERM` (https://www.freedesktop.org/software/systemd/man/systemd.service.html). The `SIGTERM` signal should indeed lead to a "clean" termination in which `vpn-slice` will be called to do the disconnect.

Are you sure you haven't configured your systemd to use `SIGKILL` (`kill -9`) instead?

**Comment by TymurGubayev** [comment]
> Are you sure you haven't configured your systemd to use `SIGKILL` (`kill -9`) instead?

I haven't changed the defaults, so the only way I'm aware of for the SIGKILL to come in play is if 90 seconds is not enough for OpenConnect to terminate - which shouldn't be the case.

Another excerpt from logs - maybe `free(): invalid pointer` followed by `Aborted` isn't clean termination.
```
/var/log/daemon.log:3272:Dec 27 10:12:11 debian-router sh[9505]: Tunnel timeout (rekey interval) is 180 minutes.
/var/log/daemon.log:3273:Dec 27 10:12:11 debian-router sh[9505]: Idle timeout is 180 minutes.
/var/log/daemon.log:3274:Dec 27 10:12:11 debian-router sh[9505]: No MTU received. Calculated 1439 for SSL tunnel. No ESP keys received
/var/log/daemon.log:3275:Dec 27 10:12:11 debian-router sh[9505]: POST https://vpn.client.com/ssl-vpn/hipreportcheck.esp
/var/log/daemon.log:3276:Dec 27 10:12:11 debian-router sh[9505]: Set up UDP failed; using SSL instead
/var/log/daemon.log:3277:Dec 27 10:12:11 debian-router sh[9505]: Connected as 123.123.123.123, using SSL, with ESP disabled
/var/log/daemon.log:3701:Dec 27 11:32:56 debian-router sh[9505]: GPST Dead Peer Detection detected dead peer!
/var/log/daemon.log:3702:Dec 27 11:32:56 debian-router sh[9505]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
/var/log/daemon.log:3703:Dec 27 11:32:56 debian-router sh[9505]: SSL negotiation with vpn.client.com
/var/log/daemon.log:3710:Dec 27 11:33:36 debian-router sh[9505]: SSL connection failure: The operation timed out
/var/log/daemon.log:3711:Dec 27 11:33:36 debian-router sh[9505]: Failed to open HTTPS connection to vpn.client.com
/var/log/daemon.log:3712:Dec 27 11:33:36 debian-router sh[9505]: sleep 10s, remaining timeout 300s
/var/log/daemon.log:3713:Dec 27 11:33:46 debian-router sh[9505]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
/var/log/daemon.log:3714:Dec 27 11:33:47 debian-router sh[9505]: SSL negotiation with vpn.client.com
/var/log/daemon.log:3715:Dec 27 11:33:47 debian-router sh[9505]: Server certificate verify failed: signer not found
/var/log/daemon.log:3716:Dec 27 11:33:48 debian-router sh[9505]: Connected to HTTPS on vpn.client.com
/var/log/daemon.log:3717:Dec 27 11:33:48 debian-router sh[9505]: Failed to parse server response
/var/log/daemon.log:3718:Dec 27 11:33:48 debian-router sh[9505]: free(): invalid pointer
/var/log/daemon.log:3723:Dec 27 11:33:48 debian-router sh[9505]: Aborted
```

**Comment by TymurGubayev** [comment]
I wrote a script looking at number of rules in `iptables` every 10 seconds, and catched the moment new rule was added.

This is how `journalctl -r -u vpn.client.service` looks like around that point of time.

```
Dec 28 15:47:04 debian-router sh[21566]: Connected as 192.168.123.123, using SSL, with ESP disabled
Dec 28 15:47:04 debian-router sh[21566]: Set up UDP failed; using SSL instead
Dec 28 15:47:04 debian-router sh[21566]: POST https://vpn.client.com/ssl-vpn/hipreportcheck.esp
Dec 28 15:47:04 debian-router sh[21566]: No MTU received. Calculated 1439 for SSL tunnel. No ESP keys received
Dec 28 15:47:04 debian-router sh[21566]: Idle timeout is 180 minutes.
Dec 28 15:47:04 debian-router sh[21566]: Tunnel timeout (rekey interval) is 180 minutes.
Dec 28 15:47:04 debian-router sh[21566]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 28 15:47:04 debian-router sh[21566]: GlobalProtect login returned authentication-source=LDAP-Client_VPN
Dec 28 15:47:04 debian-router sh[21566]: POST https://vpn.client.com/ssl-vpn/login.esp
Dec 28 15:47:04 debian-router sh[21566]: Enter login credentials
Dec 28 15:47:04 debian-router sh[21566]: Connected to HTTPS on vpn.client.com
Dec 28 15:47:03 debian-router sh[21566]: Server certificate verify failed: signer not found
Dec 28 15:47:03 debian-router sh[21566]: SSL negotiation with vpn.client.com
Dec 28 15:47:03 debian-router sh[21566]: Connected to 1.2.3.4
Dec 28 15:47:03 debian-router sh[21566]: POST https://vpn.client.com/ssl-vpn/prelogin.esp?tmp=tmp&clientVer=4100&clientos=Linux
Dec 28 15:47:03 debian-router sudo[21568]: pam_unix(sudo:session): session opened for user root by (uid=0)
Dec 28 15:47:03 debian-router sudo[21568]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/sbin/openconnect --passwd-on-stdin -s vpn-slice
Dec 28 15:47:03 debian-router systemd[1]: Started Connect to Client VPN.
Dec 28 15:47:03 debian-router systemd[1]: Stopped Connect to Client VPN.
Dec 28 15:47:03 debian-router systemd[1]: vpn.client.service: Scheduled restart job, restart counter is at 165.
Dec 28 15:47:03 debian-router systemd[1]: vpn.client.service: Service RestartSec=1min expired, scheduling restart.
Dec 28 15:46:02 debian-router systemd[1]: vpn.client.service: Failed with result 'exit-code'.
Dec 28 15:46:02 debian-router systemd[1]: vpn.client.service: Main process exited, code=exited, status=134/n/a
Dec 28 15:46:02 debian-router sh[9612]: Aborted
Dec 28 15:46:02 debian-router sudo[9614]: pam_unix(sudo:session): session closed for user root
Dec 28 15:46:02 debian-router sh[9612]: free(): invalid pointer
Dec 28 15:46:02 debian-router sh[9612]: Failed to open HTTPS connection to vpn.client.com
Dec 28 15:46:02 debian-router sh[9612]: SSL connection failure: The operation timed out
Dec 28 15:45:22 debian-router sh[9612]: SSL negotiation with vpn.client.com
Dec 28 15:45:22 debian-router sh[9612]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 28 15:45:12 debian-router sh[9612]: sleep 10s, remaining timeout 300s
Dec 28 15:45:12 debian-router sh[9612]: Failed to open HTTPS connection to vpn.client.com
Dec 28 15:45:12 debian-router sh[9612]: SSL connection failure: The operation timed out
Dec 28 15:44:32 debian-router sh[9612]: SSL negotiation with vpn.client.com
Dec 28 15:44:32 debian-router sh[9612]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 28 15:44:32 debian-router sh[9612]: Packet receive error: Input/output error
Dec 28 15:44:32 debian-router sh[9612]: SSL read error: Error in the pull function.; reconnecting.
Dec 28 15:09:04 debian-router sh[9612]: POST https://vpn.client.com/ssl-vpn/hipreportcheck.esp
Dec 28 15:09:04 debian-router sh[9612]: No MTU received. Calculated 1439 for SSL tunnel. ESP disabled
Dec 28 15:09:04 debian-router sh[9612]: Idle timeout is 180 minutes.
Dec 28 15:09:04 debian-router sh[9612]: Tunnel timeout (rekey interval) is 180 minutes.
Dec 28 15:09:04 debian-router sh[9612]: Connected to HTTPS on vpn.client.com
```

**Comment by dlenski** [comment]
> ```
> Dec 28 15:46:02 debian-router sh[9612]: free(): invalid pointer
> Dec 28 15:46:02 debian-router sh[9612]: Failed to open HTTPS connection to vpn.client.com
> Dec 28 15:46:02 debian-router sh[9612]: SSL connection failure: The operation timed out
> Dec 28 15:45:22 debian-router sh[9612]: SSL negotiation with vpn.client.com
> Dec 28 15:45:22 debian-router sh[9612]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
> ```

Are you able to verify that this apparent memory access error is coming from OpenConnect itself?

If so… could you add `--dump-http-traffic -vvv` to the OpenConnect CLI? This'll give a whole ton of extra debugging output, which should help pinpoint where the error is actually occurring.

Also, [we fixed a few memory access errors between v8.02 and v8.10](https://gitlab.com/openconnect/openconnect/-/compare/v8.02...v8.10), so it's possible that this has already been fixed in a newer version of OpenConnect.

**Comment by TymurGubayev** [comment]
I can't really verify where the error is coming from, only strongly guess from what I see in logs -- it looks indeed like a problem in OpenConnect somewhere.

Adding `--dump-http-traffic -vvv` the service journal around time duplicate iptables entry appeared looks like this.
```
Dec 29 09:15:37 debian-router sh[27061]: Attempting to connect to server 1.2.3.4
Dec 29 09:15:37 debian-router sh[27061]: POST https://vpn.client.com/ssl-vpn/prelogin.esp?tmp=tmp&clientVer=4100&clientos=Linux
Dec 29 09:15:37 debian-router systemd[1]: Started Connect to Client VPN.
Dec 29 09:15:37 debian-router sudo[27063]: pam_unix(sudo:session): session opened for user root by (uid=0)
Dec 29 09:15:37 debian-router systemd[1]: Stopped Connect to Client VPN.
Dec 29 09:15:37 debian-router sudo[27063]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/sbin/openconnect --dump-http-traffic -vvv --pass
Dec 29 09:15:37 debian-router systemd[1]: vpn.client.service: Scheduled restart job, restart counter is at 1.
Dec 29 09:15:37 debian-router systemd[1]: vpn.client.service: Service RestartSec=1min expired, scheduling restart.
Dec 29 09:14:37 debian-router systemd[1]: vpn.client.service: Failed with result 'exit-code'.
Dec 29 09:14:37 debian-router systemd[1]: vpn.client.service: Main process exited, code=exited, status=134/n/a
Dec 29 09:14:37 debian-router sh[22496]: Aborted
Dec 29 09:14:37 debian-router sudo[22498]: pam_unix(sudo:session): session closed for user root
Dec 29 09:14:37 debian-router sh[22496]: free(): invalid pointer
Dec 29 09:14:37 debian-router sh[22496]: Failed to open HTTPS connection to vpn.client.com
Dec 29 09:14:37 debian-router sh[22496]: SSL connection failure: The operation timed out
Dec 29 09:13:57 debian-router sh[22496]: SSL negotiation with vpn.client.com
Dec 29 09:13:57 debian-router sh[22496]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 29 09:13:46 debian-router sh[22496]: sleep 10s, remaining timeout 300s
Dec 29 09:13:46 debian-router sh[22496]: Failed to open HTTPS connection to vpn.client.com
Dec 29 09:13:46 debian-router sh[22496]: SSL connection failure: The operation timed out
Dec 29 09:13:06 debian-router sh[22496]: SSL negotiation with vpn.client.com
Dec 29 09:13:06 debian-router sh[22496]: POST https://vpn.client.com/ssl-vpn/getconfig.esp
Dec 29 09:13:06 debian-router sh[22496]: GlobalProtect rekey due
Dec 29 09:12:57 debian-router sh[22496]: No work to do; sleeping for 9000 ms...
Dec 29 09:12:57 debian-router sh[22496]: Got GPST DPD/keepalive response
```

Thanks for the hint about newer version of openconnect - I'll give it a try.

**Comment by TymurGubayev** [comment]
It's been almost a week since I upgraded to `OpenConnect version v8.10-1~bpo10+1`, and there have been no occurrences of this problem.

**Comment by dlenski** [comment]
That's good. Seems like this must've been due to a memory leak that was fixed subsequently to v8.02. :+1:
